packages/ui-components/src/components/flow-canvas/canvas-context.tsx

```diff
@@ -103,17 +104,163 @@ export const InteractiveContextProvider = ({
   const selectedFlowActionRef = useRef<Action | null>(null);
   const selectedNodeCounterRef = useRef<number>(0);
   const state = useStoreApi().getState();
+  const [actionToPaste, setActionToPaste] = useState<Action | null>(null);
 
   const spacePressed = useKeyPress(SPACE_KEY);
   const shiftPressed = useKeyPress(SHIFT_KEY);
 
-  const canvas = useMemo(() => {
-    return flowCanvasContainerId
-      ? document.getElementById(flowCanvasContainerId)
-      : null;
-  }, [flowCanvasContainerId]);
-  const copyPressed = useKeyPress(COPY_KEYS, { target: canvas });
-  const { fetchClipboardOperations } = useClipboardContext();
+  const fallbackCopy = (
+    text: string,
+    action: Action | null,
+    actionCount?: number,
+  ) => {
+    const textarea = document.createElement('textarea');
+    textarea.value = text;
+    textarea.style.position = 'fixed';
+    textarea.style.opacity = '0';
+    document.body.appendChild(textarea);
+    textarea.select();
+
+    try {
+      if (action) {
+        setActionToPaste(action);
+      }
+      // eslint-disable-next-line
+      if (document.execCommand) {
+        // eslint-disable-next-line
+        document.execCommand('copy');
+      }
+      copyPasteToast({
+        success: true,
+        isCopy: true,
+        itemsCount: actionCount,
+      });
+    } catch (err) {
+      copyPasteToast({
+        success: false,
+        isCopy: true,
+      });
+    } finally {
+      document.body.removeChild(textarea);
+    }
+  };
+
+  const handleCopy = useCallback(
+    (action: Action, actionCount: number) => {
+      const flowString = JSON.stringify(action);
+
+      if (navigator.clipboard) {
+        navigator.clipboard
+          .writeText(flowString)
+          .then(() => {
+            setActionToPaste(action);
+            copyPasteToast({
+              success: true,
+              isCopy: true,
+              itemsCount: actionCount,
+            });
+          })
+          .catch(() => {
+            copyPasteToast({
+              success: false,
+              isCopy: true,
+            });
+          });
+      } else {
+        if (isSafari) {
+          fallbackCopy('-', null, 0);
+          return;
+        }
+
+        fallbackCopy(flowString, action, actionCount);
+      }
```

Comment:
for some reason fallback copy the copy doesn't work in safari the first time, so had to add another hack for it